File: 08-tries/00212-Word_Search_II/00212-word_search_ii.go
// 212: Word Search Ii
// https://leetcode.com/problems/word-search-ii/

package main

import (
	"fmt"
)


type TrieNode struct{
	val rune
    word string
    next map[rune]*TrieNode
}


func dfs(board [][]byte, res *[]string, node *TrieNode, i,j int, visited map[string]bool){
    k := fmt.Sprintf("%v_%v", i,j)

    if i<0 || j<0 || i>=len(board) || j>=len(board[0]) || visited[k] { return }
    nn, ok:= node.next[rune(board[i][j])]
    if !ok {return}
    
    if len(nn.word)>0 {
        *res = append(*res, nn.word)
        nn.word = ""
    }
    
    visited[k] = true
    dfs(board, res, nn, i+1, j, visited)
    dfs(board, res, nn, i-1, j, visited)
    dfs(board, res, nn, i, j+1, visited)
    dfs(board, res, nn, i, j-1, visited)
    visited[k] = false
}


// SOLUTION
func findWords(board [][]byte, words []string) []string {
    node := &TrieNode{next : map[rune]*TrieNode{}}
    for _, w := range words {
        p := node
        for i, b := range w {
            if _, ok := p.next[b]; !ok {
                p.next[b] = &TrieNode{val:b, next:map[rune]*TrieNode{}} 
            }
            if i==len(w)-1 {p.next[b].word = w}
            p = p.next[b]
        }
    }

    result := []string{}
    for i:=0; i<len(board); i++ {
        for j:=0; j<len(board[i]); j++ {
            dfs(board, &result, node, i,j, map[string]bool{})
        }
    }
    
    return result
}


func main() {
    // INPUT
    board := [][]byte{
        {'o','a','a','n'},
        {'e','t','a','e'},
        {'i','h','k','r'},
        {'i','f','l','v'},
    }
    words := []string{"oath","pea","eat","rain"}

    // OUTPUT
    fmt.Println(findWords(board, words))
}

File: 08-tries/00212-Word_Search_II/00212-word_search_ii.java
// 212: Word Search Ii
// https://leetcode.com/problems/word-search-ii/


import java.util.ArrayList;
import java.util.List;


class Solution {
    class TrieNode {
        TrieNode[] next = new TrieNode[26];
        String word;
    }
    
    public TrieNode buildTrie(String[] words) {
        TrieNode root = new TrieNode();

        for (String w : words) {
            TrieNode p = root;
            for (char c : w.toCharArray()) {
                int i = c - 'a';
                if (p.next[i] == null) p.next[i] = new TrieNode();
                p = p.next[i];
           }
           p.word = w;
        }

        return root;
    }

    public void dfs(char[][] board, int i, int j, TrieNode p, List<String> result) {
        char c = board[i][j];
        if (c == '#' || p.next[c - 'a'] == null) return;

        p = p.next[c - 'a'];
        if (p.word != null) {
            result.add(p.word);
            p.word = null;
        }

        board[i][j] = '#';
        if (i > 0) dfs(board, i - 1, j ,p, result); 
        if (j > 0) dfs(board, i, j - 1, p, result);
        if (i < board.length - 1) dfs(board, i + 1, j, p, result); 
        if (j < board[0].length - 1) dfs(board, i, j + 1, p, result); 
        board[i][j] = c;
    }


    // SOLUTION
    public List<String> findWords(char[][] board, String[] words) {
        List<String> result = new ArrayList<>();
        TrieNode root = buildTrie(words);

        for (int i=0; i<board.length; i++) {
            for (int j = 0; j<board[0].length; j++) {
                dfs(board, i, j, root, result);
            }
        }

        return result;
    }


    public static void main(String[] args) {
        Solution o = new Solution();

        // INPUT
        char[][] board = {
            {'o','a','a','n'},
            {'e','t','a','e'},
            {'i','h','k','r'},
            {'i','f','l','v'}
        };
        String[] words = {"oath","pea","eat","rain"};

        // OUTPUT
        var result = o.findWords(board, words);
        System.out.print("["); for (var x : result) System.out.print(x+", "); System.out.println("\b\b]");
    }
}
